### ISSUE TITLE
How to pass props using inline?

### ISSUE THREAD

**Issue by zaydek** [nonzzz/stylex-extend#11]
I'm trying to do something like the following:

```tsx
function Foo({ foo }: { foo: any }) {
  return (
    <div {...stylex.props(foo)}>
      helloasdasdasd
    </div>
  );
}

export function App() {
  return (
    <>
      <Foo foo={inline({ backgroundColor: "red" })} />
      <div {...stylex.props(inline({ color: "blue" }))}>
        hello
      </div>
    </>
  );
}
```

Is it possible to create your own prop for accepting `inline` on a component?

But I get this error:

```
react-dom_client.js?v=857fe929:19441 Uncaught Error: 'inline' calls should be compiled away.
    at inline (@stylex-extend_core.js?t=1716341981246&v=857fe929:8:9)
    at App (app.tsx:37:1)
    at renderWithHooks (react-dom_client.js?v=857fe929:11568:26)
    at mountIndeterminateComponent (react-dom_client.js?v=857fe929:14946:21)
    at beginWork (react-dom_client.js?v=857fe929:15934:22)
    at beginWork$1 (react-dom_client.js?v=857fe929:19781:22)
    at performUnitOfWork (react-dom_client.js?v=857fe929:19226:20)
    at workLoopSync (react-dom_client.js?v=857fe929:19165:13)
    at renderRootSync (react-dom_client.js?v=857fe929:19144:15)
    at recoverFromConcurrentError (react-dom_client.js?v=857fe929:18764:28)
```
    
Let me know if you need a repro.

**Comment by zaydek** [comment]
```tsx
import { inline } from "@stylex-extend/core";
import * as stylex from "@stylexjs/stylex";

const red = inline({ backgroundColor: "red" });

function Foo({ extend }: { extend?: any }) {
  return (
    <div {...stylex.props(extend)}>
      hello
    </div>
  );
}

export function App() {
  return (
    <>
      <Foo extend={red} />
    </>
  );
}
```

```
@stylex-extend_core.js?t=1716342266224&v=857fe929:8 Uncaught Error: 'inline' calls should be compiled away.
    at inline (@stylex-extend_core.js?t=1716342266224&v=857fe929:8:9)
    at app.tsx:5:3
```

I can't quite tell if `extend`'s type is supposed to be `stylex.StyleXStyles` or something else.

**Comment by nonzzz** [comment]
Oh, Currently i made a restriction on inline api. It only can called with as a argument with function call. Like this code block can be process

```ts
import { inline } from "@stylex-extend/core";

const fn = <T>(...args: T[]) => {
  return args;
};

// correct
const result = fn(inline({ color: "red" }));

// incorrect

const result2 = inline({ color: "pink" });
```

Follow the correct case it will be translate as

```ts
import { create as _create, props as _props } from "@stylexjs/stylex";
const _styles = _create({
  "#0": {
    color: "red",
  },
});
import { inline } from "@stylex-extend/core";
const fn = (...args) => {
  return args;
};
const result = fn(_styles["#0"]);
```

So i think you can try define a wrapper function to implment the extend feature.

**Comment by zaydek** [comment]
Yes that worked — thank you! I'll use the `fn` trick for now. Feel free to ping me if you have an update on this.

**Comment by nonzzz** [comment]
Restricting inline calls is intended to prevent the abuse of this behavior.

**Comment by zaydek** [comment]
But wouldn't it be ideal to say

```tsx
<Foo extend={atom({ backgroundColor: "red" })} />
```

Instead of

```tsx
<Foo extend={fn(atom({ backgroundColor: "red" }))} />
```

?

**Comment by nonzzz** [comment]
Right. using a function wrapper it.

**Comment by nonzzz** [comment]
About this limit i think i should find a balance to design it.  As you can see is too hard to use when implementing extend.

**Comment by zaydek** [comment]
Take your time. I'm going to experiment with https://github.com/facebook/stylex/issues/534#issuecomment-2121745213 in the meantime.

**Comment by predaytor** [comment]
really cool! but using this wrapper unfortunately doesn't work:
```ts
import { inline } from '@stylex-extend/core';
import type { CSSObject } from '@stylex-extend/shared';

const fn = <T>(...args: T[]) => args;

export const atom = (args: CSSObject) => fn(inline(args));
```
only direct definition works:
```ts
export const atom = (args: CSSObject) => fn(inline({ opacity: '50%' }));
```

**Comment by nonzzz** [comment]
@predaytor It's a limit.

**Comment by piotrblasiak** [comment]
Any plans to remove this limitation? Also, as mentioned before the return type is not compatible with StyleXStyles?

**Comment by zaydek** [comment]
> Any plans to remove this limitation? Also, as mentioned before the return type is not compatible with StyleXStyles?

For what it's worth, I found a happy middle ground to be something like the following: Use inline styles to figure out the structure/nature of what you are building first, and then, once you feel like you have a good grip on it, you can quickly convert it to the stylesheet API. If you do that often enough, you will feel more comfortable skipping the inline styles step, but it's a good fallback if you are feeling blocked.

This package helps, but you can simply author inline styles in StyleX, and it will technically work (you might get TypeScript errors that you can subvert). However, StyleX already propagates inline styles: `<div {...stylex.props({ /* Inline styles */ })}>` should work today with only the minimum StyleX integration.

The only caution I can offer is that if you start relying on macros or plugins like this one, there is some risk that switching to a different Vite/plugin integration could cause issues in the future.

**Comment by piotrblasiak** [comment]
> > Any plans to remove this limitation? Also, as mentioned before the return type is not compatible with StyleXStyles?
> 
> For what it's worth, I found a happy middle ground to be something like the following: Use inline styles to figure out the structure/nature of what you are building first, and then, once you feel like you have a good grip on it, you can quickly convert it to the stylesheet API. If you do that often enough, you will feel more comfortable skipping the inline styles step, but it's a good fallback if you are feeling blocked.
> 
> This package helps, but you can simply author inline styles in StyleX, and it will technically work (you might get TypeScript errors that you can subvert). However, StyleX already propagates inline styles: `<div {...stylex.props({ /* Inline styles */ })}>` should work today with only the minimum StyleX integration.
> 
> The only caution I can offer is that if you start relying on macros or plugins like this one, there is some risk that switching to a different Vite/plugin integration could cause issues in the future.

Sure maybe in time but converting a huge codebase from emotion for example becomes a huge burden. And I doubt I will ever be happy with having to come up with names for every single element that needs a style.

**Comment by zaydek** [comment]
> Sure maybe in time but converting a huge codebase from emotion for example becomes a huge burden. And I doubt I will ever be happy with having to come up with names for every single element that needs a style.

Fair enough. The nuance with Emotion (if it's modern Emotion) is that it's very easy to accidentally write deoptimized code if you're not careful. It's basically impossible to write deoptimized StyleX, but it's also more work upfront. The only real sin you can commit with StyleX is using too many CSS variables, but there's no runtime context overhead or anything like that that I'm aware of.  

I found that it's sometimes easier to work if I simply have the common styles ready to go. You can prepare a couple of utility styles that you can consume anywhere; that way, you are not getting in your own way with things that feel like they ought to be pure utilities. For example: `styles.displayFlex`, `styles.justifyContentCenter`, etc. You probably only need a handful, and then if you feel you want to add more than a few properties, just extract it into its own style object instead. I've also used components, but they are a little harder to reason about. When I first started using StyleX, this is all I cared about and explored it pretty exhaustively here: https://github.com/facebook/stylex/issues/534#issuecomment-2121745213.  

Long story short, my findings have been that StyleX will challenge you in the way you think about authoring styles, but it will make you a better engineer because it forces you to recognize patterns and name things (taxonomy), which serves as a form of documentation to your future self. I will say that while libraries like Emotion and Tailwind are generally more productive at the beginning, the maintenance burden becomes really unwieldy. I have found that maintaining StyleX code to be really straightforward and predictable. 

If you have other thoughts or questions feel free to follow up, I've used StyleX on quite a few projects now and I feel like I understand its strengths and weaknesses pretty well now.

**Comment by nonzzz** [comment]
Fine, About type error, i think @mengdaoshizhongxinyang  will help me :D and it's time to lift the inline limit. postcss integrate for vite will come soon, (base on official plugin)

**Comment by nonzzz** [comment]
Secondly, macro will evaluate static values as well as possible ( or fallback to official stylex engine) in my product env, works well until now.

**Comment by piotrblasiak** [comment]
Amazing, looking forward to it - especially now that stylex also have removed stylex.include in 0.11 😣

**Comment by nonzzz** [comment]
I released version 0.7.0. I removed the restriction and provided another bundle with vite.
